// vecks/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "V2", {
  enumerable: true,
  get: function get() {
    return _V["default"];
  }
});
Object.defineProperty(exports, "V3", {
  enumerable: true,
  get: function get() {
    return _V2["default"];
  }
});
Object.defineProperty(exports, "Box2", {
  enumerable: true,
  get: function get() {
    return _Box["default"];
  }
});
Object.defineProperty(exports, "Box3", {
  enumerable: true,
  get: function get() {
    return _Box2["default"];
  }
});
Object.defineProperty(exports, "Plane3", {
  enumerable: true,
  get: function get() {
    return _Plane["default"];
  }
});
Object.defineProperty(exports, "Quaternion", {
  enumerable: true,
  get: function get() {
    return _Quaternion["default"];
  }
});
Object.defineProperty(exports, "Line2", {
  enumerable: true,
  get: function get() {
    return _Line["default"];
  }
});
Object.defineProperty(exports, "Line3", {
  enumerable: true,
  get: function get() {
    return _Line2["default"];
  }
});

var _V = _interopRequireDefault(require("./V2"));

var _V2 = _interopRequireDefault(require("./V3"));

var _Box = _interopRequireDefault(require("./Box2"));

var _Box2 = _interopRequireDefault(require("./Box3"));

var _Plane = _interopRequireDefault(require("./Plane3"));

var _Quaternion = _interopRequireDefault(require("./Quaternion"));

var _Line = _interopRequireDefault(require("./Line2"));

var _Line2 = _interopRequireDefault(require("./Line3"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
